text/0000-move-pointer.md

```diff
@@ -0,0 +1,103 @@
+- Feature Name: `move_pointer`
+- Start Date: 2016-05-12
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Introduce a new pointer type `&move` that logically owns the pointed-to data, but does not control the backing memory. Also introduces the DerefMove trait to provide access to such a pointer.
+
+# Motivation
+[motivation]: #motivation
+
+This provides an elegant solution to passing DSTs by value, allowing `Box<FnOnce>` to work. It also will allow other usecases where trait methods should take self by "value".
+
+# Detailed design
+[design]: #detailed-design
+
+- Add a new pointer type `&move T`
+ - This pointer will get the same lifetime rules as `&` and `&mut` (such that it cannot outlive the allocation)
+ - but, it will allow the value to become invalid before the allocation does.
+ - Deref coercions as per RFC #241 apply.
+ - Variables of type `&move T` will only allow mutating `T` if they themselves are `mut`
+- Add a new unary operation `&move <value>`. With a special case such that `&move |x| x` parses as a move closure, requiring parentheses to parse as an owned pointer to a closure.
+ - This precedence can be implemented by ignoring the `move` when parsing unary operators if it is followed by a `|` or `||` token.
+- Add a new operator trait `DerefMove` to allow smart pointer types to return owned pointers to their contained data. 
+ - A type that implements `DerefMove` cannot implement `Drop` (as `DerefMove` provides equivalent functinality)
+
+```rust
+trait DerefMove: DerefMut
+{
+    /// Return an owned pointer to inner data
+    fn deref_move(&mut self) -> &move Self::Target;
```

Comment:
And finally, dropping, which is vary easy:
```rust
impl<T> Drop for Box<T>
{
    fn drop_interior(arg: &move Interior<Box<T>>) {
        let x: &move T = arg.0;
        drop(reborrow_move(x));
        // x: OutPtr<T> --- we can still use it
        let ptr: *u8 = x as _;
        unsafe {
            heap::deallocate(ptr, mem::size_of::<T>(), mem::align_of::<T>());
        }
    }
}
```
Dropping `Uninit<T>` is a no-op, so we don't need to worry about dropping `DeadBox`es at all!